### PR TITLE
Ensure default variation is carried over to Facebook

### PR DESCRIPTION
**Pull request by N-Molham**
# Summary

This PR will ensure when the default variation changes on the website it syncs (carry over) to Facebook store

### Setup

- Make sure product sync is enabled and all products are synced
- Create a product with two or more variations

### Steps

1. Edit the variable product and change it's default `Default Form Values` to something corresponds to matching variation
    - [ ] Changes get saved normally without error
    - [ ] It reflects on the Shop section in the Facebook page 

image

image

## Before merge

- [ ] I have confirmed these changes in each supported minor WooCommerce version

### CODE REVIEW

File: facebook-commerce.php
Comment (unfulvio) [inline facebook-commerce.php:1436]: if I am not mistaken in old (yet supported) versions of WooCommerce (see for example WC 3.0.x) this method would only return an array and has no support for `objects` - in this case the foreach loop below would produce a PHP error; we should ensure this not to be a possibility and watch out for types

**Review by unfulvio — CHANGES_REQUESTED** [review]
hey @N-Molham in principle this works but the code needs some changes to account for old WC versions -- thanks

File: facebook-commerce.php
Comment (N-Molham) [reply facebook-commerce.php:1427]: Yes, because in runtime the product data type passed to the parent calling method `WC_Facebookcommerce_Integration::update_product_group` is by the same data type of `\WC_Facebook_Product` ... so to keep the workflow the same without breaking it, I just continued with the same data type

Comment (N-Molham) [reply facebook-commerce.php:1427]: Maybe I just used to make methods like that `protected` by default to avoid changing it down the road in case the the class got extended ... I just need to change this habit

File: facebook-commerce.php
Comment (N-Molham) [reply facebook-commerce.php]: this 648ea1f should take care of that

**Review by unfulvio — APPROVED** [review]
that looks better @N-Molham - given that 2.1.0 is released though (see https://github.com/facebookincubator/facebook-for-woocommerce/releases/tag/2.1.0), we  need to rebase this

**Comment by N-Molham** [comment]
@unfulvio I assume you want to rabase it on 2.1.1 release, right?

**Comment by wvega** [comment]
@N-Molham we deployed 2.1.1 not too long ago to address high impact bug, so we will need to target this for another release. 

We are trying to get all the changes from 2.1.0 and 2.1.1 merged into master so that we can have clean release branches again, so I would say we wait until #1655 is merged before we merging this one into a new release branch (which doesn't exist yet).

**Comment by wvega** [comment]
@unfulvio @N-Molham I just rebased this branch on top of #1658. For reference, the previous set of commits was https://github.com/facebookincubator/facebook-for-woocommerce/compare/c88c9e76...8a937163

File: facebook-commerce.php
Comment (wvega) [inline facebook-commerce.php:1473]:
```suggestion
	 * Determines if there is a matching variation for the default attributes.
```

File: changelog.txt
Comment (wvega) [inline changelog.txt:4]: Since the product item ID of the default variation is not available on initial sync (because each variation will be synced asynchronously using the Batch API), I think we should update the changelog entry to indicate when can merchants expect to see the default variation reflected on their catalogs.

Perhaps something like this (should update `readme.txt` as well):

```suggestion
 * Tweak - Default variation selection will be synced to Facebook if the default product variation is already synced 

```

File: facebook-commerce.php
Comment (wvega) [inline facebook-commerce.php:1441]: @N-Molham looks like `$woo_default_variation` is an array now but `get_product_fbid()` expects a `WC_Facebook_Product` in the third argument. Perhaps we don't need to pass the third argument and should allow `get_product_fbid()` to create the instance instead?

```suggestion
				$woo_default_variation['variation_id']
```

**Review by wvega — CHANGES_REQUESTED** [review]
@N-Molham Changes are looking good!

I took a look at this one trying to prepare the 2.1.2 patch release and left a couple of suggestions to update the changelog entry and fix the parameters to `get_product_fbid()`.

File: facebook-commerce.php
Comment (N-Molham) [reply facebook-commerce.php:1441]: Yeah, you are right ... the `get_product_fbid()` take care of that if the 3rd argument is `null`

**Comment by N-Molham** [comment]
@wvega Took care of your comments in 13d3f5d & 0119d3b

**Review by wvega — APPROVED** [review]
@N-Molham changes look good. Thank you for the updates. I'll merge with a final tweak to update the changelog entry in the `readme.txt` file as well: b2f3097

**Pull request merged**
